spec/abci++/abci++_app_requirements_002_draft.md

```diff
@@ -177,3 +179,818 @@ Likewise, `ExtendVote` can also be non-deterministic:
+has run a successful `Commit` so that it can tell Tendermint where to
+pick up from when it restarts. See information on the Handshake, below.
+
+#### Commit
+
+The Application state should be persisted during `Commit`.
+
+Before invoking `Commit`, Tendermint locks the mempool and flushes the mempool connection. This ensures that
+no new messages
+will be received on the mempool connection during this processing step, providing a good opportunity to safely update all four
+connection states to the latest committed state at once.
+
+When `Commit` completes, Tendermint unlocks the mempool.
+
+WARNING: if the ABCI app logic processing the `Commit` message sends a
+`/broadcast_tx_sync` or `/broadcast_tx_commit` and waits for the response
+before proceeding, it will deadlock. Executing `broadcast_tx` calls
+involves acquiring the mempool lock that Tendermint holds during the `Commit` call.
+If you make the call to the `broadcast_tx` endpoints concurrently,
+that's not a problem, it just can't be part of the sequential logic of the
+`Commit` function.
+
+#### Candidate States
+
+Tendermint calls `PrepareProposal` when it is about to send a proposed block to the network.
+Likewise, Tendermint calls `ProcessProposal` upon reception of a proposed block from the
+network. In both cases, the proposed block's data (last block's commit info, transactions, etc.)
+is disclosed to the Application, in the same conditions as is done in `FinalizeBlock`.
+
+The Application may decide to *immediately* execute the given block (i.e., upon `PrepareProposal`
+or `ProcessProposal`). There are two main reasons why the Application may want to do this:
+
+* In order to be sure that the block does not contain *any* invalid transaction, there is usually
+  no way other than fully executing the transactions in the block as though it was the *decided*
+  block.
+* Upon reception of the decided block via `FinalizeBlock`, if that same block was executed
+  upon `PrepareProposal` or `ProcessProposal` and the resulting state was kept in memory, the
+  Application can simply apply that state (faster) to the main state, rather than reexecuting
+  the decided block (slower).
+
+`PrepareProposal`/`ProcessProposal` can be called many times for a given height. Moreover,
+it is not possible to accurately predict which of the blocks proposed in a height will be decided, being delivered to the application
+in that height's `FinalizeBlock`.
+Therefore, the state resulting from executing a proposed block, denoted a *candidate state*, should
+be kept in memory as a possible final state for that height. When `FinalizeBlock` is called, the Application should
+check if the decided block corresponds to one of its candidate states; if so, it will apply it as
+its `ExecuteTxState` (see [Consensus Connection](#consensus-connection) below),
+which will be persisted during the upcoming `Commit` call.
+
+In bad runs (e.g., under network instability), Consensus might take many rounds.
```

Comment:
I would opt for something that's more clear than "bad runs". 

I'm also worried about the the phrasing of "Consensus" as an actor that is part of the specification (rather than just saying "Tendermint"? I'm not sure that people need to think about our internal state machine.

Reply:
OK, reworded